[ChenXy128] iP

ChenXy128 opened the pull request:
# DukeJinni
> Jinni is a Duke in 1800 AC
> 
Your **latest** assistant to help you sort things out! Jinni is:

- Free 😇
- Fast
- Reliable
- EASY to use! 

### All you need to do is,
1. Download it [here ]()
2. Open it
3. Start adding your tasks

### Jinni manages:

- [ ] _tasks_
- [X] _deadlines_
- [ ] _events_

If you want to learn Java, here is the `main` class for you to start Jinni

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

ChenXy128 committed:
L0 done

ChenXy128 committed:
L1 done

ChenXy128 committed:
L2 done

ChenXy128 committed:
Task class

ChenXy128 committed:
L3 Done

ChenXy128 committed:
L4 Done

ChenXy128 committed:
Deadlines

ChenXy128 committed:
Events

ChenXy128 committed:
Todos

ChenXy128 committed:
L5 Done

ChenXy128 committed:
DukeException

ChenXy128 committed:
L6 Done

ChenXy128 committed:
Enum Done

ChenXy128 committed:
Enums editied

ChenXy128 committed:
Expected Output

ChenXy128 committed:
User Input

ChenXy128 committed:
Formatted the indentation of switch-case block by left-shifting 4 whitespaces

ChenXy128 committed:
Enabled saving to hardisk

ChenXy128 committed:
Added parseStringToTask method

ChenXy128 committed:
Merge branch 'branch-Level-7'

ChenXy128 committed:
Changed the by attribute from String to LocalDate

ChenXy128 committed:
Changed the start and end attributes from String to LocalDate

ChenXy128 committed:
Added method formatDateForParsing, formatDateForPrinting and parseFileDateToStorageDate methods

ChenXy128 committed:
Modified Deadline and Event commands

ChenXy128 committed:
Merge branch 'branch-Level-8'

ChenXy128 committed:
A more OOP Duke file

ChenXy128 committed:
Parser class that deals with understanding user input

ChenXy128 committed:
Storage class

ChenXy128 committed:
Updated task

ChenXy128 committed:
Tasklist class to deal with the tasks

ChenXy128 committed:
User Interaction class that deals with interacting with the user

ChenXy128 committed:
Merge branch 'A-MoreOOP'

ChenXy128 committed:
A-Packages

ChenXy128 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

ChenXy128 committed:
Updated Storage

ChenXy128 committed:
Add ParserTest.java
Add StorageTest.java

ParserTest.java contains 2 methods
* createToDo_validInput_success() tests for createToDo method correctly creates a ToDo object if the user input is correct
* createToDo_invalidInput_throwsException() tests for a wrong user input will not be accepted.

ChenXy128 committed:
Modify Ui.java
Change the corresponding code in Duke.java

Delete the extra space between "in" and "the" in the deleting method.
Remove the unused argument inputFromUser in handleDeleteError method
Remove the argument in run method that calls handleDeleteError

ChenXy128 committed:
Add JavaDoc for all non-private methods in Duke.java, Parser.java, Storage.java and Ui.java

ChenXy128 committed:
Adjust the code to follow the coding standard

Things changed:
* rename class ToDos to ToDo
* rename class Deadlines to Deadline
* rename class Events to Event
* correct the imports of the classes with names modified in other classes
* correct the codes when creating objects of the classes with names modified

ChenXy128 committed:
Implement query for tasks method with the keyword

Methods added:
* Duke has a new case FIND to deal with the new instruction find
* Parser has a new method getKeywordForFind to retrieve the keyword from user input
* TaskList has a findTask method, which is the main logic of the find function
* Ui has a new finding method to interact with the user when they are finding for tasks

ChenXy128 committed:
Add JavaDoc

JavaDoc for the files are updated and added.

ChenXy128 committed:
Modify Duke.java's run() method

In case FIND
* Before: this.ui.finding(duke.list, keyword)
* After: duke.ui.finding(duke.list, keyword)

This is to be consistent such that I am using the ui object from the duke object created, instead of the ui from Duke's attribute

wapisai commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,204 @@
+import java.io.IOException;
+import java.io.File;
+
+/**
+ * Represents the main class of the Duke application, a personal assistant chatbot
+ * that helps users to keep track of various tasks. The application is capable of
+ * understanding commands to add, delete, mark, and list tasks among other functionalities.
+ *
+ * Duke uses a combination of components to manage tasks:
+ * - {@link duke.storage.Storage} for loading from and saving tasks to a file.
+ * - {@link duke.parser.Parser} for interpreting user input.
+ * - {@link duke.ui.Ui} for handling interactions with the user.
+ * - {@link duke.tasklist.TaskList} for storing and managing a list of tasks.
+ *
+ * The application supports tasks of types: Todo, Deadline, and Event, which can be
+ * added, marked as done, deleted, and listed. It operates in a loop, processing user
+ * commands until a termination command is received.
+ */
```
Very elaborate! Good job!

wapisai commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,204 @@
+        duke.ui.greeting();
+    
+        while(isRunning) {
+            String inputFromUser = sc.nextLine();
```
perhaps "userInput" rather than "inputFromUser"

wapisai commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,204 @@
+                break;
+            case MARK:
+                duke.ui.handleMarkError(inputFromUser, duke.list);
+                Task taskToBeMarked = duke.parser.getTaskTobeMarked(inputFromUser, duke.list);
```
Maybe "taskToMark"

wapisai commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,204 @@
+                int indexOfTaskToDelete = Integer.parseInt(inputFromUser.substring(7));
+                duke.ui.handleDeleteError(duke.list,indexOfTaskToDelete);
+                Task taskToDelete = duke.parser.getTaskToDelete(inputFromUser, duke.list);
+                duke.ui.deleting(taskToDelete, duke.list);
```
could be more specific on what this method does, since its ui maybe use "printDeleteMessage"

wapisai commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,204 @@
+                duke.ui.handleDeleteError(duke.list,indexOfTaskToDelete);
+                Task taskToDelete = duke.parser.getTaskToDelete(inputFromUser, duke.list);
+                duke.ui.deleting(taskToDelete, duke.list);
+                duke.list.delete(taskToDelete);
```
this could be "deleteTask" just to be a little more specific especially when put in close proximity with your ui method "deleting"

wapisai commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,204 @@
+                break;
+            case FIND:
+                String keyword = duke.parser.getKeywordForFind(inputFromUser);
+                duke.ui.finding(duke.list, keyword);
```
Like previously, perhaps have a more specific method name like printFoundList

wapisai commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,204 @@
+                duke.ui.finding(duke.list, keyword);
+                break;
+            case BYE:
+                duke.ui.bye();
```
Could perhaps use "printOutro"

wapisai commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,177 @@
+     * @return The Task to be deleted.
+     */
+    public Task getTaskToDelete(String input, TaskList list) {
+        int indexOfTaskToDelete = Integer.parseInt(input.substring(7)) - 1;
```
I think here could write `int i` instead of that since its used in a small scope?

wapisai commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,106 @@
+     * @param list The TaskList containing all current tasks.
+     * @throws IOException If an I/O error occurs during writing to the file.
+     */
+    public void changeFileContent(TaskList list) throws IOException {
```
maybe "updateFileContent"

wapisai reviewed (commented):
Overall a good job! However, naming for methods in UI class could be improved!

Derekljh commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,204 @@
+     * @param input The raw input string from the user.
+     * @return The Instruction that corresponds to the user's command.
+     */
+    public Instruction getInstr(String input) {
+        if (input.toLowerCase().startsWith("list")) {
+            return Instruction.LIST;
+        } else if (input.toLowerCase().startsWith("mark")) {
+            return Instruction.MARK;
+        } else if (input.toLowerCase().startsWith("unmark")) {
+            return Instruction.UNMARK;
+        } else if (input.toLowerCase().startsWith("todo")) {
+            return Instruction.TODO;
+        } else if (input.toLowerCase().startsWith("deadline")) {
+            return Instruction.DEADLINE;
+        } else if (input.toLowerCase().startsWith("event")) {
+            return Instruction.EVENT;
+        } else if (input.toLowerCase().startsWith("delete")) {
+            return Instruction.DELETE;
+        } else if (input.toLowerCase().startsWith("find")) {
+            return Instruction.FIND;
+        } else if (input.toLowerCase().startsWith("bye")){
+            return Instruction.BYE;
+        } else {
+            return Instruction.ANYTHING_ELSE;
+        }
+    }
```
Proper use of K&R style brackets. Good job!

Derekljh commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,204 @@
+            String inputFromUser = sc.nextLine();
+            Instruction instruction = duke.getInstr(inputFromUser);
+        
+            switch (instruction) {
+            case LIST:
```
Following coding standards, no indentation for case clauses. Well done!

Derekljh commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,177 @@
+     * @return A new Event Task.
+     * @throws DukeException If any input date is invalid or the input format is incorrect.
+     */
+    public Task createEvent(String input, Ui ui) throws DukeException{
+        int indexOfFrom = input.indexOf("/from");
+        int indexOfTo = input.indexOf("/to");
+        String eventDescription = input.substring(6, indexOfFrom - 1);
+        String startString = input.substring(indexOfFrom + 6, indexOfTo - 1);
+        String endString = input.substring(indexOfTo + 4);
+        LocalDate start = this.formatInputDate(startString);
+        LocalDate end = this.formatInputDate(endString);
+        ui.handleInvalidInputDate(startString, DateTimeFormatter.ofPattern("dd/MM/yyyy"), start);
+        ui.handleInvalidInputDate(endString, DateTimeFormatter.ofPattern("dd/MM/yyyy"), end);
+        return new Event(eventDescription, start, end);
+    }
```
Maybe can separate this chunk of code with a few blank lines to enhance readability?

Derekljh commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,177 @@
+     * @param line The string representation of the task from the storage file.
+     * @return The corresponding Task object.
+     */
+    public Task parseStringToTask(String line) {
+        char typeOfTask = line.charAt(1);
+        char status = line.charAt(4);
+        Task t = null;
+        if (typeOfTask == 'T') {
+            String description = line.substring(7);
+            t = new ToDo(description);
+        } else if (typeOfTask == 'D') {
+            int indexOfTime = line.indexOf("(by: ");
+            String description = line.substring(7, indexOfTime - 1);
+            String byString = line.substring(indexOfTime + 5, line.indexOf(")"));
+            LocalDate by = parseFileDateToStorageDate(byString);
+            t = new Deadline(description, by);
+        } else if (typeOfTask == 'E') {
+            int indexOfStartTime = line.indexOf("(from");
+            int indexOfEndTime = line .indexOf("to");
+            String description = line.substring(7, indexOfStartTime - 1);
+            String startString = line.substring(indexOfStartTime + 6, indexOfEndTime - 1);
+            String endString = line.substring(indexOfEndTime + 3, line.indexOf(")"));
+            LocalDate start = parseFileDateToStorageDate(startString);
+            LocalDate end = parseFileDateToStorageDate(endString);
+            t = new Event(description, start, end);
+        }
+        boolean doneOrNot = (status == 'X');
+        t.setStatus(doneOrNot);
+        return t;
+    }
```
Maybe can do the same as before, separate some of the lines inside this method with blank lines.

Derekljh commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,235 @@
+        if (list.size() == 0) {
+            System.out.println("\n___________________________________");
+        } else {
+            int num = 1;
+            for (Task t : list.getList()) {
```
Nice that you initialized the variables where they are declared, and in the smallest scope possible.

Derekljh reviewed (commented):
Overall, coding standards were all met.
Looks good to me!

ChenXy128 committed:
Update README.md

Created a new description of my ip according to the specifications.

ChenXy128 committed:
Enable to run the application on gradle and update javadoc

Created new files:
* DialogBox.fxml
* DialogBox.java
* Launcher.java
* Main.java
* MainWindow.fxml
* MainWindow.java
* User.png
* Jinni.png

Updated file:
* build.gradle
* Duke.java
* Parser.java
* User.java

updated JavaDoc for all files

ChenXy128 committed:
Modified the classes so that the chatbot will not cause any truncation

ChenXy128 committed:
Removes unused import statement

The import from package Ui has been removed from Parser.java

ChenXy128 committed:
Adds assert to ensure that important assumptions hold

String response should not be null at all times in getResponse() method in Duke.java
Task t should not be null at all times in parseStringToTask method in Parser.java

ChenXy128 committed:
Shortens getResponse method

A Command.java class is created to handle the logic of getResponse of various commands

ChenXy128 committed:
Adds JavaDoc to class Command.java

ChenXy128 committed:
Implements extension C-Help

ChenXy128 committed:
Merge pull request #2 from ChenXy128/branch-BCD-Extension

Implements extension C-Help

ChenXy128 committed:
Improves method getResponse by further shortening it

ChenXy128 committed:
Merge pull request #3 from ChenXy128/branch-A-CodeQuality

Improves method getResponse by further shortening it

ChenXy128 committed:
Adds JavaDoc to almost all public methods

ChenXy128 committed:
Merge pull request #4 from ChenXy128/branch-A-JavaDoc

Adds JavaDoc to almost all public methods

ChenXy128 committed:
Update README.md

ChenXy128 committed:
Adds a Ui.png, updates README.md

ChenXy128 committed:
Merge remote-tracking branch 'origin/master'

ChenXy128 committed:
Update README.md

ChenXy128 committed:
Update README.md

ChenXy128 committed:
Update README.md

ChenXy128 committed:
Update README.md

ChenXy128 committed:
Updates file path and build a fat jar
